### PR TITLE
SUMO to AWS data pipeline

### PR DESCRIPTION
**Pull request by pgrandinetti**
SUMO to AWS pipeline

Run with
/bin/bash std_traffic/pipelines/sumoToAws.sh \
    SUMO_COMMAND='-c /home/pietro/Desktop/MoSTScenario/scenario/most.sumocfg -e 15000 --fcd-output' \
    SUMO_OUTPUT_FILE='saveout' \
    SUMO_MODEL_PREFIX='most' \
    SUMO_TO_CSV='/home/pietro/Desktop/sumo/tools/xml2csv.py' \
    S3=0 \
    DB=1 \
    DATABASE=sumodb \
    TABLE=test_csv \
    ENVR=/home/pietro/Desktop/standard-traffic-data/.env

To do:
- I/O for pipeline script
- Docs
- Timeout tests

### CODE REVIEW

**Comment by pgrandinetti** [comment]
I will want to merge this sometime early next week. Send me comments, if any, by then.

File: std_traffic/utils/aws_utils.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        aws_access_key_id=environ['AWS_KEY_ID'],
+        aws_secret_access_key=environ['AWS_SECRET_KEY'],
+        region_name=environ['AWS_REGION'])
+
+
+def get_s3_client() -> boto3.client:
+    """Create a connection client to S3."""
+
+    return _init_client('s3')
+
```
Comment (RuggeroFabbiano) [inline std_traffic/utils/aws_utils.py:26]: Why this and not use directly boto3.client? Security?

File: std_traffic/utils/data_utils.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+) -> connection:
+    """Create a connection to a database.
+    The default connection parameters are taken from ENV
+    """
+
+    if not user:
+        user = environ['DB_USER']
+    if not password:
+        password = environ['DB_PASSWORD']
+    if not host:
```
Comment (RuggeroFabbiano) [inline std_traffic/utils/data_utils.py:34]: Couldn't it go directly in the default parameter definition?

File: scripts/load_csv_psql.py [hunk 3]
```diff
@@ -41,28 +42,35 @@
...
         default=5432)
-
     args = parser.parse_args()
+
+    if not args.host:
+        args.host = environ['DB_HOST']
+    if not args.user:
+        args.user = environ['DB_USER']
+    if not args.password:
+        args.password = environ['DB_PASSWORD']
```
Comment (RuggeroFabbiano) [inline scripts/load_csv_psql.py]: Why this change?

File: std_traffic/utils/aws_utils.py [hunk 4]
```diff
@@ -8,26 +8,30 @@
...
-        aws_access_key_id=environ['AWS_KEY_ID'],
-        aws_secret_access_key=environ['AWS_SECRET_KEY'],
-        region_name=environ['AWS_REGION'])
+        aws_access_key_id=aws_access_key_id,
+        aws_secret_access_key=aws_secret_access_key,
+        region_name=region_name)
 
 
-def get_s3_client() -> boto3.client:
+def get_s3_client(**kwargs) -> boto3.client:
```
Comment (RuggeroFabbiano) [inline std_traffic/utils/aws_utils.py:22]: Why?

File: std_traffic/utils/data_utils.py [hunk 5]
```diff
@@ -44,29 +44,72 @@ def get_db(
...
-    all_types = dict(dataf.dtypes)
-    return all_types
+    # - Read first 1000 row, get columns names, infer types
+    #   by avoiding NA columns
+    # - For each column type missing
+    #   keep reading only that column in chunks until a type is found
+    #   or EOF
+    initial = 1000
+    dataf = pd.read_csv(csv_file, sep=delim, nrows=initial)
+    all_names = set(dataf.columns)
```
Comment (RuggeroFabbiano) [inline std_traffic/utils/data_utils.py]: Why cast it as set?

File: std_traffic/utils/aws_utils.py (hunk 1 above)
Comment (pgrandinetti) [reply std_traffic/utils/aws_utils.py:26]: @RuggeroFabbiano There are multiple clients exposed by the `boto3` package, and the only thing that changes to initialize them is the name. So, with the code above you could do

```
_init_client('s3')
```

but also

```
_init_client('sns')
```

Using `boto3.client` would require to always give the credentials (or to have them in the "famous" `/home/.awsconfig` file, but I don't like that solution).

What alternative do you recommend?

File: std_traffic/utils/data_utils.py (hunk 2 above)
Comment (pgrandinetti) [reply std_traffic/utils/data_utils.py:34]: @RuggeroFabbiano If you mean

```
def foo(arg_1, arg_2=environ[KEY_2])
```

that would work the same, but it would force the evaluation of `environ[KEY_2]` much before the function is used. It would be evaluated whenever the whole module `data_utils.py` is imported, and that feels unnatural to me.
For that reason, the import would fail with a `KeyError` if `KEY_2` doesn't exist in `environ`, even if the user meant to not use the default value for the argument.
This last problem could be solved using `getenv(KEY_2)` instead of 'envirion[...]` but it still looks odd to me in the function definition.
Is this what you meant?

File: std_traffic/pipelines/sumoToAws.sh [hunk 6]
```diff
@@ -330,40 +368,67 @@ load_db() {
...
+        logit "Exiting with error. CSV saved at ${TMP_DIR}/${SUMO_MODEL_PREFIX}.${SUMO_OUTPUT_FILE}.csv"
+    fi
+
+    if [ $CSV_SPLIT_DONE -eq 0 ]
+    then rm $TMP_DIR/${SUMO_MODEL_PREFIX}.${SUMO_OUTPUT_FILE}.csv.0*
+    fi
 fi
 
-logit "All done"
+logit "Done."
```
Comment (RuggeroFabbiano) [inline std_traffic/pipelines/sumoToAws.sh:423]: Don't know if I got it right, but here you're basically continuing all the way down to the end of the script even if a problem arised and completing the log even if doing anything, instead of just stopping the execution? If so, why?

File: std_traffic/pipelines/sumoToAws.sh [hunk 7]
```diff
@@ -412,21 +404,19 @@ fi
...
         logit "Exiting with error. CSV saved at ${TMP_DIR}/${SUMO_MODEL_PREFIX}.${SUMO_OUTPUT_FILE}.csv"
     fi
 
     if [ $CSV_SPLIT_DONE -eq 0 ]
-    then rm $TMP_DIR/${SUMO_MODEL_PREFIX}.${SUMO_OUTPUT_FILE}.csv.0*
+    then
+        logit "Removing CSV chunks"
+        rm $TMP_DIR/${SUMO_MODEL_PREFIX}.${SUMO_OUTPUT_FILE}.csv.0*
     fi
 fi
```
Comment (RuggeroFabbiano) [inline std_traffic/pipelines/sumoToAws.sh:422]: OK I guess the answer to my previous comment is: it was temporary :)
I leave it anyway so you can confirm if I understood.

**Review by RuggeroFabbiano — APPROVED** [review]
Vista la quantità di lavoro che hai prodotto non ho guardato ogni linea nel dettaglio per non farti perdere altro tempo con una cosa che sarebbe stata più utile a me che a te.

Nelle grandi linee quello che hai fatto mi è chiaro (e ovviamente mi sembra fatto molto bene :)). Ho fatto comunque dei commenti, che saranno probabilmente più utili a me che a te... il grosso mi sembra riguardi il perché hai trattato gli argomenti degli script con degli _if_ invece che impostarli di _default_, e forse un paio di altre cose.

Ho scelto apposta di riguardare tutto in ordine cronologico per capire il filo logico che avevi seguito, quindi molti commenti che ho eliminato si son auto-chiariti lungo la strada; ne ho lasciato uno di questi lo stesso per esser sicuro di aver capito.

File: std_traffic/utils/aws_utils.py (hunk 1 above)
Comment (RuggeroFabbiano) [reply std_traffic/utils/aws_utils.py:26]: I was meaning, in `_upload_file` you call `client = get_s3_client()`, that calls `_init_client('s3')`, that calls `boto3.client(args)`;
why not call in `client = boto3.client(args)` directly?

File: std_traffic/utils/data_utils.py (hunk 2 above)
Comment (RuggeroFabbiano) [reply std_traffic/utils/data_utils.py:34]: Yes it is, and I see it now.

File: scripts/load_csv_psql.py (hunk 3 above)
Comment (pgrandinetti) [reply scripts/load_csv_psql.py]: @RuggeroFabbiano this conversation doesn't show where those arguments were before my change, so I can't respond fully. I remember I had initially set them as `default` in the `ArgumentParser` and then I got error because it tried to evaluate `environ[...]` before checking if the default was needed or not. A bit like the other comment of yours above.

File: std_traffic/utils/aws_utils.py (hunk 1 above)
Comment (pgrandinetti) [reply std_traffic/utils/aws_utils.py:26]: The idea of `get_s3_client` is that it doesn't need any input and therefore has more control over what you get from it.
In another situation, I would tell you that the module hides the implementation of the `_init_client` because the user is not supposed to change that. Of course, in Python this is trash talking.
I still think it's nice anyway not to have to pass all the `args` around.
Do you have a suggestion to improve this part?

File: std_traffic/utils/data_utils.py (hunk 5 above)
Comment (pgrandinetti) [reply std_traffic/utils/data_utils.py]: @RuggeroFabbiano Search operation (like `in` in Python) is O(1) with sets and O(n) with lists, like `dataf.columns` would be.
However, looking at the rest of the function, this cast is basically useless because the search in the set is done in one place only. And anyway the size of list of always relatively small (it's the number of features, after all).
It was more of an instinctive reflex to convert it to `set` because I knew I was going to search in it.
Do you think we should remove it?

File: std_traffic/pipelines/sumoToAws.sh (hunk 6 above)
Comment (pgrandinetti) [reply std_traffic/pipelines/sumoToAws.sh:423]: @RuggeroFabbiano You did, but this part went through a lot of changes because I wasn't sure what files make more sense to delete in case of failure, and which one not.
Sure we could stop anytime and just `exit 1`, and add some `rm` there, but after all I thought it was OK to have the cleanup part at the bottom.
How would you change it?

File: std_traffic/utils/aws_utils.py (hunk 1 above)
Comment (RuggeroFabbiano) [reply std_traffic/utils/aws_utils.py:26]: Yes I guess I see the point (some sort of "fake encapsulation" if I get it right), but since `get_s3_client()` is only used in an already "private method" it seems to me that the implementation is already hidden in some sense, and all these levels can be avoided. It would be different if it was "public". Moreover, this would avoid creating a "public" `get_s3_client()` that can be used by the user (thus making the need to protect `_init_client()`). Is that wrong?

File: std_traffic/utils/data_utils.py (hunk 5 above)
Comment (RuggeroFabbiano) [reply std_traffic/utils/data_utils.py]: Since the search is done only once I'd say "yes", given that the conversion is also _O_(_n_).

File: std_traffic/pipelines/sumoToAws.sh (hunk 6 above)
Comment (RuggeroFabbiano) [reply std_traffic/pipelines/sumoToAws.sh:423]: On second thought yes, I guess it's neater that way.

File: std_traffic/utils/aws_utils.py (hunk 1 above)
Comment (pgrandinetti) [reply std_traffic/utils/aws_utils.py:26]: I think you are suggesting to remove `_init_client(...)` and have instead that code inside `get_s3_client()`, like this:

```
def get_s3_client(**kwargs) -> boto3.client:
    aws_access_key_id = kwargs.get('AWS_KEY_ID', environ['AWS_KEY_ID'])
    aws_secret_access_key = kwargs.get('AWS_SECRET_KEY', environ['AWS_SECRET_KEY'])
    region_name = kwargs.get('AWS_REGION', environ['AWS_REGION'])
    return boto3.client(
        's3',
        aws_access_key_id=aws_access_key_id,
        aws_secret_access_key=aws_secret_access_key,
        region_name=region_name)
```

I don't like this because then, one day, I will have to write a `get_sns_client(...)` function, for a different boto3 client service, and I will have to rewrite the same 8 lines, just changing the first argument of the last call to `'sns'`.
Whereas, with the intermediate function `_init_client` I could just do

```
def get_sns_client(**kwargs):
    return _init_client('sns', **kwargs)
```

Maybe the trailing underscore of `_init_client` is confusing: my point here is more about avoiding duplicate code, much less about function visibility.

File: std_traffic/utils/data_utils.py (hunk 5 above)
Comment (pgrandinetti) [reply std_traffic/utils/data_utils.py]: Fixed in 1756eae

File: std_traffic/utils/aws_utils.py (hunk 1 above)
Comment (RuggeroFabbiano) [reply std_traffic/utils/aws_utils.py:26]: I would more keep just `_init_client()` and use it directly into `_update_file()`, so one day you can just call `_init_client()` with `'sns'`.

So just to summarise, `_init_client()` helps to DRY code, whereas `get_s3_client()` is used to "hide" what you're doing underneath, right?
